### ISSUE TITLE
Enhancement: [require-await] add an option to ignore async generators or support yield Promise

### ISSUE THREAD

**Issue by ziloen** [typescript-eslint/typescript-eslint#7995]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/require-await/

### Description

According to https://github.com/eslint/eslint/issues/17790#issuecomment-1826277508, ESLint ignores async generators, maybe add an option to ignore async generators or support yield Promise?

### Fail

```typescript
async function* test() {
  yield 1
  yield 2
}
```

### Pass

```typescript
async function* test2() {
  yield new Promise(() => {})
}
```

**Comment by bradzacher** [comment]
ESLint's core rule also reports on `async () => Promise.resolve()` - but we don't because our rule is type-aware and thus knows that the return value is a promise. This is one of the few instances where mirroring their exact behaviour is not necessarily an intended goal.

Back when we added support for async generators in #1782 the case of yielding a promise wasn't handled. We support generators that `await`, and generators that delegate to async generators.

I don't know enough about generators to know the correct behaviour here - @Josh-Cena / @JoshuaKGoldberg do either of you know enough to weigh in?

Throwing the following into the chrome console:
```ts
async function* test1() {
  yield Promise.resolve(1);
}
(async () => {
  for await (const x of test1()) {
    console.log(x);
  }
})();

function* test2() {
  yield Promise.resolve(2);
}
(async () => {
  for await (const x of test2()) {
    console.log(x);
  }
})();
```

Makes it look like this case might be similar to the `return <promise>` case - but I don't know really anything about generators.

**Comment by Josh-Cena** [comment]
Each `yield` in an asynchronous generator also implicitly unwraps promises just like `return`, so it should definitely be handled like `return`.
